includes/class-wc-gateway-swedbank-pay-cc.php

```diff
@@ -338,63 +424,34 @@ public function validate_fields() {
 	public function add_payment_method() {
 		$user_id = get_current_user_id();
 
-		// Get Customer UUID
-		if ( $user_id > 0 ) {
-			$customer_uuid = get_user_meta( $user_id, '_payex_customer_uuid', true );
-			if ( empty( $customer_uuid ) ) {
-				$customer_uuid = swedbank_pay_uuid( $user_id );
-				update_user_meta( $user_id, '_payex_customer_uuid', $customer_uuid );
-			}
-		} else {
-			$customer_uuid = swedbank_pay_uuid( uniqid( 'add_payment_method' ) );
-		}
-
-		$params = [
-			'payment' => [
-				'operation'               => 'Verify',
-				'currency'                => get_woocommerce_currency(),
-				'description'             => __( 'Verification of Credit Card', WC_Swedbank_Pay::TEXT_DOMAIN ),
-				'payerReference'          => $customer_uuid,
-				'generatePaymentToken'    => true,
-				'generateRecurrenceToken' => true,
-				'pageStripdown'           => false,
-				'userAgent'               => isset( $_SERVER['HTTP_USER_AGENT'] ) ? $_SERVER['HTTP_USER_AGENT'] : '',
-				'language'                => $this->culture,
-				'urls'                    => [
-					'completeUrl' => add_query_arg( 'action', 'swedbank_card_store', admin_url( 'admin-ajax.php' ) ),
-					'cancelUrl'   => wc_get_account_endpoint_url( 'payment-methods' ),
-					'callbackUrl' => WC()->api_request_url( __CLASS__ )
-				],
-				'payeeInfo'               => [
-					'payeeId'        => $this->payee_id,
-					'payeeReference' => swedbank_pay_uuid( uniqid( 'add_payment_method' ) ),
-				],
-				'creditCard'              => $this->get_card_options()
-			]
-		];
-
-		// Add subsite
-		if ( ! empty( $this->subsite ) ) {
-			$params['payment']['payeeInfo']['subsite'] = $this->subsite;
-		}
+		// Create a virtual order
+		$order = wc_create_order(
+			array(
+				'customer_id'    => $user_id,
+				'created_via'    => $this->id,
+				'payment_method' => $this->id,
+			)
+		);
+		$order->calculate_totals();
 
 		try {
-			$result = $this->request( 'POST', '/psp/creditcard/payments', $params );
+			$this->is_new_credit_card = true;
+			$result                   = $this->core->initiateNewCreditCardPayment( $order->get_id() );
 		} catch ( Exception $e ) {
-			$this->log( sprintf( '[ERROR] Process payment: %s', $e->getMessage() ) );
+			wc_add_notice( $e->getMessage(), 'error' );
 
 			WC()->session->__unset( 'verification_payment_id' );
 
-			return [
+			return array(
 				'result'   => 'failure',
 				'redirect' => wc_get_account_endpoint_url( 'payment-methods' ),
-			];
+			);
 		}
 
 		WC()->session->set( 'verification_payment_id', $result['payment']['id'] );
 
 		// Redirect
-		wp_redirect( self::get_operation( $result['operations'], 'redirect-verification' ) );
+		wp_redirect( $result->getOperationByRel( 'redirect-verification' ) );
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [wp_redirect() found. Using wp_safe_redirect(), along with the allowed_redirect_hosts filter if needed, can help avoid any chances of malicious redirects within code. It is also important to remember to call exit() after a redirect so that no other unwanted code is executed.](https://app.codacy.com/gh/SwedbankPay/swedbank-pay-woocommerce-payments/pullRequest?prid=5340032)